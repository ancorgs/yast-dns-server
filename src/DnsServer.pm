#! /usr/bin/perl -w
#
# DnsServer module written in Perl
#

package DnsServer;

use strict;

use ycp;
use YaST::YCP qw(Boolean);
use Data::Dumper;
use Time::localtime;

#use io_routines;
#use check_routines;

our %TYPEINFO;

# persistent variables

my $start_service = 0;

my $chroot = 0;

my @allowed_interfaces = ();

my @zones = ();

my %options = ();

my %logging = ();

#transient variables

my $modified = 0;

my $save_all = 0;

my @files_to_delete = ();

my %current_zone = ();

my $current_zone_index = -1;

my $adapt_firewall = 0;

my %firewall_settings = ();

my $write_only = 0;


# FIXME remove this func

sub _ {
    return $_[0];
}


YaST::YCP::Import ("SCR");
YaST::YCP::Import ("Mode");
YaST::YCP::Import ("Service");

##-------------------------------------------------------------------------
##----------------- various routines --------------------------------------

BEGIN { $TYPEINFO{AbsoluteZoneFileName} = ["function", "string", "string" ]; }
sub AbsoluteZoneFileName {
    my $file_name = $_[0];

    if (substr ($file_name, 0, 1) eq "/")
    {
	return $file_name;
    }
    return "/var/lib/named/$file_name";
}

sub UpdateSerial {
    my $serial = $_[0];

    if (! defined ($serial))
    {
	$serial = "0000000000";
    }

    my $year = 1900 + localtime->year();
    my $month = 1 + localtime->mon();
    my $day = localtime->mday();

    while (length ($month) < 2)
    {
	$month = "0$month";
    }
    while (length ($day) < 2)
    {
	$day = "0$day";
    }
    while (length ($year) > 4)
    {
	$year = substr ($year, 1);
    }
    my $date = "$year$month$day";
    my $suffix = "00";
    if (substr ($serial, 0, 8) eq $date)
    {
	$suffix = substr ($serial, 8, 2);
	$suffix = $suffix + 1;
	while (length ($suffix) < 2)
	{
	    $suffix = "0$suffix";
	}
	while (length ($suffix) > 2)
	{
	    $suffix = substr ($suffix, 1);
	}
    }
    $serial = "$date$suffix";
    y2milestone ("New serial $serial");
    $serial;
}

sub contains {
    my @list = @{$_[0]};
    my $value = $_[1];

    my $found = 0;
    foreach my $x (@list) {
	if ($x eq $value)
	{
	    $found = 1;
	    last;
	}
    }
    $found;
}

##------------------------------------
# Routines for reading/writing configuration

BEGIN { $TYPEINFO{ZoneRead} = ["function", [ "map", "any", "any" ], "string", "string" ]; }
sub ZoneRead {
    my $zone = $_[0];
    my $file = $_[1];

    my %zonemap = %{SCR::Read (".dns.zone", "/var/lib/named/$file")};
    my %soa = %{$zonemap{"soa"} || {}};
    $soa{"serial"} = UpdateSerial ($soa{"serial"} || "");
    my %ret = (
	"zone" => $zone,
	"ttl" => $zonemap{"TTL"} || "2W",
	"soa" => \%soa,
    );
    my @original_records = @{$zonemap{"records"}};
    my %in_mx = ();
    my %in_prt = ();
    my %in_cname = ();
    my %in_a = ();
    my $previous_key = "$zone.";

    my %records = ();
    foreach my $r (@original_records) {
	my %r = %{$r};
	my $key = $r{"key"} || "";
	my $type = $r{"type"} || "";
	my $value = $r{"value"} || "";

	if ($key eq "")
	{
	    $key = $previous_key;
	}
	else
	{
	    $previous_key = $key;
	}
	my %host = %{$records{$key} || {}};
	my @items = @{$host{$type} || []};
	push (@items, $value);
	$host{$type} = \@items;
	$records{$key} = \%host;
    }

    $ret{"records"} = \%records;

    return %ret;
}

BEGIN { $TYPEINFO{ZoneFileUpdate} = ["function", "boolean", [ "map", "any", "any" ]];}
sub ZoneFileUpdate {
    my %zone_map = %{$_[0]};
    my @actions = @{$zone_map{"update_actions"}};
    foreach my $action (@actions) {
	
	# TODO perform the action
    }
}

BEGIN { $TYPEINFO{ZoneFileWrite} = ["function", "boolean", [ "map", "any", "any"]];}
sub ZoneFileWrite {
    my %zone_map = %{$_[0]};

    my $zone_file = $zone_map{"file"} || "";
    $zone_file = AbsoluteZoneFileName ($zone_file);
    my $zone_name = $zone_map{"zone"} || "@";
    my $ttl = $zone_map{"ttl"} || "2W";

    my %soa = (
	"zone" => "@",
	"expiry" => "6W",
	"mail" => "root",
	"minimum" => "1W",
	"refresh" => "2D",
	"retry" => "4H",
	"server" => "@",
    );
    my %current_soa = %{$zone_map{"soa"}};
    while ((my $key, my $value) = each %current_soa)
    {
	$soa{$key} = $value;
    }

    my @records = ();

    my %records = %{$zone_map{"records"} || {}};
    while ((my $key, my $values_ref) = each (%records)) {
	my %values = %{$values_ref};
	my @types = keys (%values);
	my @preferred = ("A", "CNAME", "PTR");
	@preferred = grep {
	    contains ( \@preferred, $_);
	} @types;
	my @others = grep {
	    ! contains (\@preferred, $_);
	} @types;
	@types = ( @preferred, @others );
	foreach my $type (@types) {
	    my $res_rec_ref = $values{$type};
	    my @res_rec = @{$res_rec_ref};
	    foreach my $rr (@res_rec) {
		my %new_rec = (
		    "key" => $key,
		    "type" => $type,
		    "value" => $rr,
		);
		push (@records, \%new_rec);
	    }
	}
    }

    my %save = (
	"TTL" => $ttl,
	"soa" => \%soa,
	"records" => \@records,
    );
    return SCR::Write (".dns.zone", [$zone_file, \%save]);
}

BEGIN { $TYPEINFO{ZoneWrite} = ["function", "boolean", [ "map", "any", "any" ] ]; }
sub ZoneWrite {
    my %zone_map = %{$_[0]};

    if (! ($zone_map{"modified"} || $save_all))
    {
	return 1;
    }

    my $zone_name = $zone_map{"zone"} || "";
    if ($zone_name eq "")
    {
	return 0;
    }

    my $zone_file = $zone_map{"file"} || "";
    if ($zone_file eq "")
    {
	$zone_file = "master/$zone_name";
	$zone_map{"file"} = $zone_file;
    }

    #save changed of named.conf
    my $base_path = ".dns.named.value.\"zone \\\"$zone_name\\\" in\"";
    SCR::Write (".dns.named.section.\"zone \\\"$zone_name\\\" in\"", "");

    my @old_options = SCR::Dir ($base_path) || ();
    my @save_options = @old_options;

    my $zone_type = $zone_map{"type"} || "master";

    if ($zone_type eq "master")
    {
	# write the zone file
	if (! $zone_map{"soa_modified"})
	{
	    ZoneFileWrite (\%zone_map);
	}
	else
	{
	    ZoneFileUpdate (\%zone_map);
	}
	@save_options = ("type", "file");

	# write existing keys
	SCR::Write ("$base_path.$zone_file", "\"$zone_file\"");
    }
    elsif ($zone_type eq "slave" || $zone_type eq "stub")
    {
	@save_options = ("masters");
	my $masters = $zone_map{"masters"} || "";
	if (! $masters =~ /\{.*;\}/)
	{
	    $zone_map{"masters"} = "{$masters;}";
	}
        SCR::Write ("$base_path.masters", $zone_map{"masters"} || "");
    }
    elsif ($zone_type eq "hint")
    {
	@save_options = ("type", "file");
	SCR::Write ("$base_path.file", "\"$zone_file\"");
    }

    my @del_options = grep {
	! contains (\@save_options, $_);
    } @old_options;
    foreach my $o (@del_options) {
	SCR::Write ("$base_path.$o", undef);
    };
    SCR::Write ("$base_path.type", $zone_map{"type"} || "master");

    return 1;
}

BEGIN { $TYPEINFO{AdaptFirewall} = ["function", "boolean"]; }
sub AdaptFirewall {
    if (! $adapt_firewall)
    {
	return 1;
    }

    SCR::Write (".sysconfig.SuSEfirewall2.FW_SERVICE_DNS", $start_service
	? "yes"
	: "no");

# TODO enable or disable firewall for particular interfaces

    SCR::Write (".sysconfig.SuSEfirewall2", undef);
    if (! $write_only)
    {
	SCR::Execute (".target.bash", "test -x /sbin/rcSuSEfirewall2 && /sbin/rcSuSEfirewall2 status && /sbin/rcSuSEfirewall2 restart");
    }
}

BEGIN { $TYPEINFO{SaveGlobals} = [ "function", "boolean" ]; }
sub SaveGlobals {

    #delete all deleted zones first
    my @old_sections = SCR::Dir (".dns.named.section") || ();
    my @old_zones = grep (/^zone/, @old_sections);
    my @current_zones = map {
	my %zone = %{$_};
	"zone \"$zone{\"zone\"}\" in";
    } @zones;
    my @del_zones = grep {
	! contains (\@zones, $_);
    } @old_zones;
    y2milestone ("Deleting zones @del_zones");
    foreach my $z (@del_zones) {
	SCR::Write (".dns.named.section.$z", undef);
    }

    # delete all removed options
    my @old_options = SCR::Dir (".dns.named.value.options") || ();
    my @current_options = keys (%options);
    my @del_options = grep {
	! contains (\@current_options, $_);
    } @old_options;
    foreach my $o (@del_options) {
	SCR::Write (".dns.named.value.options.$o", undef);
    }

    # save the settings
    while ((my $key, my $value) = each %options)
    {
	SCR::Write (".dns.named.value.options.$key", $value);
    }

    # really save the file
    return SCR::Write (".dns.named", undef);
}



##------------------------------------
# Store/Find/Select/Remove a zone

BEGIN { $TYPEINFO{StoreZone} = ["function", "boolean"]; }
sub StoreZone {
    $current_zone{"modified"} = 1;
    if ($current_zone_index == -1)
    {
	push (@zones, \%current_zone);
    }
    else
    {
	$zones[$current_zone_index] = \%current_zone;
    }
}

BEGIN { $TYPEINFO{FindZone} = ["function", "integer", "string"]; }
sub FindZone {
    my $zone_name = $_[0];

    my $found_index = -1;
    my $index = -1;

    map {
	$index = $index + 1;
	my %zone_map = %{$_};
	if ($zone_map{"zone"} eq $zone_name)
	{
	    $found_index = $index;
	}
    } @zones;
    return $found_index;
}

BEGIN { $TYPEINFO{RemoveZone} = ["function", "boolean", "string", "boolean"]; }
sub RemoveZone {
    my $zone_name = $_[0];
    my $delete_file = $_[1];

    my $zone_index = FindZone ($zone_name);
    if ($zone_index == -1)
    {
	y2error ("Zone $zone_name not found");
	return 0;
    }

    if ($delete_file)
    {
	my %zone_map = %{$zones[$zone_index]};
	my $filename = AbsoluteZoneFileName ($zone_map{"file"});
	push (@files_to_delete, $filename) if (defined ($filename));
    }

    $zones[$zone_index] = 0;

    @zones = grep {
	ref ($_);
    } @zones;
    return 1;
}

BEGIN { $TYPEINFO{SelectZone} = ["function", "boolean", "integer"]; }
sub SelectZone {
    my $zone_index = $_[0];

    my $ret = 1;

    if ($zone_index < -1)
    {
	y2error ("Zone with index $zone_index doesn't exist");
	$zone_index = -1;
	$ret = 0;
    }
    elsif ($zone_index >= @zones)
    {
	y2error ("Zone with index $zone_index doesn't exist");
	$zone_index = -1;
	$ret = 0;
    }

    if ($zone_index == -1)
    {
	my $serial =  UpdateSerial ("");
	my %new_soa = (
	    "expiry" => "6W",
	    "mail" => "root",
	    "minimum" => "1W",
	    "refresh" => "2D",
	    "retry" => "4H",
	    "serial" => $serial,
	    "server" => "@",
	    "zone" => "@",
	);
	%current_zone = (
	    "soa_modified" => 1,
	    "modified" => 1,
	    "type" => "master",
	    "soa" => \%new_soa,
	    "ttl" => "2W",
	);
    }
    else
    {
	%current_zone = %{$zones[$zone_index]};
    }

    return $ret;
}

##------------------------------------
# Functions for accessing the data

BEGIN { $TYPEINFO{SetStartService} = [ "function", "void", "boolean" ];}
sub SetStartService {
    $start_service = $_[0];
    SetModified ();
}

BEGIN { $TYPEINFO{GetStartService} = [ "function", "boolean" ];}
sub GetStartService {
    return $start_service;
}

BEGIN { $TYPEINFO{SetModified} = ["function", "void" ]; }
sub SetModified {
    $modified = 1;
}

BEGIN { $TYPEINFO{WasModified} = ["function", "boolean" ]; }
sub WasModified {
    return $modified;
}

BEGIN { $TYPEINFO{SetWriteOnly} = ["function", "void", "boolean" ]; }
sub SetWriteOnly {
    $write_only = $_[0];
}

BEGIN { $TYPEINFO{SetAdaptFirewall} = ["function", "void", "boolean" ]; }
sub SetAdaptFirewall {
    $adapt_firewall = $_[0];
}

BEGIN { $TYPEINFO{GetAdaptFirewall} = [ "function", "boolean" ];}
sub GetAdaptFirewall {
    return $adapt_firewall;
}

BEGIN {$TYPEINFO{FetchCurrentZone} = [ "function", ["map", "any", "any"] ]; }
sub FetchCurrentZone {
    return \%current_zone;
}

BEGIN {$TYPEINFO{StoreCurrentZone} = [ "function", "boolean", ["map", "any", "any"] ]; }
sub StoreCurrentZone {
    %current_zone = %{$_[0]};
    return 1;
}

BEGIN {$TYPEINFO{FetchZones} = [ "function", ["list", ["map", "any", "any"] ] ]; }
sub FetchZones {
    return \@zones;
}

BEGIN {$TYPEINFO{StoreZones} = [ "function", "void", [ "list", ["map", "any", "any"] ] ]; }
sub StoreZones {
    @zones = @{$_[0]};
}

BEGIN {$TYPEINFO{GetGlobalOptions} = [ "function", ["map", "any", "any"] ]; }
sub GetGlobalOptions {
    return \%options;
}

BEGIN {$TYPEINFO{SetGlobalOptions} = [ "function", "void", [ "map", "any", "any" ] ]; }
sub SetGlobalOptions {
    %options = ${$_[0]};
}

BEGIN {$TYPEINFO{GetGlobalOption} = [ "function", "any", "any" ];}
sub GetGlobalOption {
    my $key = $_[0];

    return $options{$key};
}

BEGIN {$TYPEINFO{SetGlobalOption} = ["function", "void", "any", "any"];}
sub SetGlobalOption {
    my $key = $_[0];
    my $value = $_[1];

    $options{$key} = $value;
}

BEGIN {$TYPEINFO{RemoveGlobalOption} = ["function", "void", "any" ];}
sub RemoveGlobalOption {
    my $key = $_[0];

    delete ($options{$key});
}

##------------------------------------
BEGIN { $TYPEINFO{Read} = ["function", "boolean"]; }
sub Read {
# Check packages
# TODO

# Information about the daemon

    $start_service = Service::Enabled ("named");
    y2milestone ("Service start: $start_service");
    $chroot = SCR::Read (".sysconfig.named.NAMED_RUN_CHROOTED", "")
	? 1
	: 0;
    y2milestone ("Chroot: $chroot");

    my @zone_headers = SCR::Dir (".dns.named.section");
    @zone_headers = grep (/^zone/, @zone_headers);
    y2milestone ("Read zone headers @zone_headers");

    my @opt_names = SCR::Dir (".dns.named.value.options");
    if (! @opt_names)
    {
	@opt_names = ();
    }
    foreach my $key (@opt_names) {
	$options{$key} = SCR::Read (".dns.named.value.options.$key");
    }
    foreach my $key (keys(%options)) {
    }

    @zones = map {
	my $zonename = $_;
	$zonename =~ s/.*\"(.*)\".*/$1/;
	my $path_el = $_;
	$path_el =~ s/\"/\\\"/g;
	$path_el = "\"$path_el\"";
	my $zonetype = SCR::Read (".dns.named.value.$path_el.type");
	my $filename = SCR::Read (".dns.named.value.$path_el.file");
	if ($filename =~ /^\".*\"$/)
	{
	    $filename =~ s/^\"(.*)\"$/$1/;
	}
	my %zd = ();
	if ($zonetype eq "master")
	{
	    %zd = ZoneRead ($zonename, $filename);
	}
	elsif ($zonetype eq "slave" || $zonetype eq "stub")
	{
	    $zd{"masters"} = SCR::Read (".dns.named.value.$_.masters");
	    if ($zd{"masters"} =~ /\{.*;\}/)
	    {
		$zd{"masters"} =~ s/\{(.*);\}/$1/
	    }
	}
	else
	{
# TODO hint, forward, .... not supported at the moment
	}
	$zd{"file"} = $filename;
	$zd{"type"} = $zonetype;
	$zd{"zone"} = $zonename;
	\%zd;
    } @zone_headers;
    $modified = 0;
    return "true";
}

BEGIN { $TYPEINFO{Write} = ["function", "boolean"]; }
sub Write {
    my $ok = 1;

    if (! $modified)
    {
	return "true";
    }

    #adapt firewall
    $ok = AdaptFirewall () && $ok;

    #save globals
    $ok = SaveGlobals () && $ok;

    #save all zones
    foreach my $z (@zones) {
	$ok = ZoneWrite ($z) && $ok;
    }

    #be sure the named.conf file is saved
    SCR::Write (".dns.named", undef);
    
    #set daemon starting
    SCR::Write (".sysconfig.named.NAMED_RUN_CHROOTED", $chroot ? "yes" : "no");
    SCR::Write (".sysconfig.named", undef);

    if ($start_service)
    {
	my $ret = 0;
	if (! $write_only)
	{
	    $ret = SCR::Execute (".target.bash", "/etc/init.d/named restart");
	}
	Service::Enable ("named");
	if (0 != $ret)
	{
	    $ok = 0;
	}
    }
    else
    {
	if (! $write_only)
	{
	    SCR::Execute (".target.bash", "/etc/init.d/named stop");
	}
	Service::Disable ("named");
    }

    return $ok;
}

BEGIN { $TYPEINFO{Export}  =["function", [ "map", "any", "any" ] ]; }
sub Export {
    my %ret = (
	"start_service" => $start_service,
	"chroot" => $chroot,
	"allowed_interfaces" => \@allowed_interfaces,
	"zones" => \@zones,
	"options" => \%options,
	"logging" => \%logging,
    );
    return \%ret;
}
BEGIN { $TYPEINFO{Import} = ["function", "void", [ "map", "any", "any" ] ]; }
sub Import {
    my %settings = %{$_[0]};

    $start_service = $settings{"start_service"} || 0;
    $chroot = $settings{"chroot"} || 1;
    @allowed_interfaces = @{$settings{"allowed_interfaces"} || []};
    @zones = @{$settings{"zones"} || []}; 
    %options = %{$settings{"options"} || {}};
    %logging = %{$settings{"logging"} || {}};

    $modified = 1;
    $save_all = 1;
    @files_to_delete = ();
    %current_zone = ();
    $current_zone_index = -1;
    $adapt_firewall = 0;
    $write_only = 0;
}

BEGIN { $TYPEINFO{Summary} = ["function", [ "list", "string" ] ]; }
sub Summary {
    my %zone_types = (
	# type of zone to be used in summary
	"master" => _("master"),
	# type of zone to be used in summary
	"slave" => _("slave"),
	# type of zone to be used in summary
	"stub" => _("stub"),
	# type of zone to be used in summary
	"hint" => _("hint"),
	# type of zone to be used in summary
	"forward" => _("forward"),
    );
    my @ret = ();

    if ($start_service)
    {
	# summary string
	push (@ret, _("The DNS server starts when booting the system."));
    }
    else
    {
	push (@ret,
	    # summary string
	    _("The DNS server does not start when booting the system."));
    }

    my @zones_descr = map {
	my $zone_ref = $_;
	my %zone_map = %{$zone_ref};	
	my $zone_name = $zone_map{"zone"} || "";
	my $zone_type = $zone_map{"type"} || "";
	$zone_type = $zone_types{$zone_type} || $zone_type;
	my $descr = "";
	if ($zone_name ne "")
	{
	    if ($zone_type ne "")
	    {
		$descr = "$zone_name ($zone_type)";
	    }
	    else
	    {
		$descr = "$zone_name";
	    }
	}
	$descr;
    } @zones;
    @zones_descr = grep {
	$_ ne "";
    } @zones_descr;

    my $zones_list = join (", ", @zones_descr);
    #  summary string, $zones_list is list of DNS zones (their names)
    push (@ret, _("Configured Zones: $zones_list"));
    return @ret;
}


# EOF
