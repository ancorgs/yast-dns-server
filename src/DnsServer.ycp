/**
 * File:	modules/DnsServer.ycp
 * Package:	Configuration of dns-server
 * Summary:	Data for configuration of dns-server, input and output functions.
 * Authors:	Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 * Representation of the configuration of dns-server.
 * Input and output routines.
 */

{

    module "DnsServer";
    textdomain "dns-server";

    import "Progress";
    import "Report";
    import "Summary";
    import "Runlevel";

    include "dns-server/misc.ycp";
    include "dns-server/dialogs.ycp";
    include "dns-server/wizards.ycp";
    include "dns-server/helps.ycp";
    include "dns-server/options.ycp";
    include "dns-server/widgets.ycp";

    /**
      * Data was modified?
      */
    global boolean modified = false;

    /**
      * save everything, not only changed settings
      */
    global boolean save_all = false;

    /**
      * Were the zones modified?
      */
    global boolean zones_modified = false;

    /**
      * Write only, used during autoinstallation.
      * Don't run services and SuSEconfig, it's all done at one place.
      */
    global boolean write_only = false;

    /**
      * BIND version
      */
    global integer version = 0;

    /**
      * Is read allowed to ask user about BIND package to install
      * Probably to be used for command line iface
      */
    global boolean interactive_read = true;
    /**
      * hostname
      */
    global string hostname = "localhost";
    /**
      * Date (for zone files)
      */
    global string date = "";

    /**
      * DNS server settings
      */
    global map settings = $[];

    /**
      * Temporary list of zones
      */
    global list all_zones = [];

    /**
      * Temporary storage of current zone
      */
    global map current_zone = $[];

    /**
      * Temporary variable for storing original id of currently edited zone
      */
    global integer current_id = -1;

    global string displayed_settings = "settings";

    /**
      * Map of popups for CWM
      */
    global map popups = InitPopups ();

    /**
      * Map of widgets for CWM
      */
    global map widgets = InitWidgets ();

    /**
      * Read all dns-server settings
      * @return true on success
      */
    global define boolean Read() ``{

	/* DnsServer read dialog caption */
	string caption = _("Initializing DNS Server Configuration");

	integer steps = 30;

	Progress::New( caption, " ", steps, [
		// progress stage
		_("Check the environment"),
		// progress stage
		_("Read main configuration file"),
		// progress stage
		_("Read the zone files")
	    ], [
		// progress step
		_("Checking the environment..."),
		// progress step
		_("Reading main configuration file..."),
		// progress step
		_("Reading the zone files..."),
		// progress step
		_("Finished")
	    ],
	    ""
	);

	// read database
	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (0);

	// set the local host name
	hostname = getLocalHostName ();

	// determine used BIND version
	version = getBindVersion (interactive_read);
	if (version == nil)
	{
	    // warning report
	    Report::Warning (_("Neither bind8 nor bind9 installed"));
	}

	settings["start_service"] = Runlevel::ServiceEnabled ("named");
	settings["chroot"]
	    = SCR::Read (.sysconfig.named.NAMED_RUN_CHROOTED) == "yes"
		? true
		: false;

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (10);

	list zone_headers = SCR::Dir (.dns.named.section);
	zone_headers = filter (`z, zone_headers,
	    ``(substring (z, 0, 4) == "zone"));
	y2milestone ("Read zone sections %1", zone_headers);

	list opt_names = SCR::Dir (.dns.named.value.options);
	if (opt_names == nil)
	    opt_names = [];
	map options = listmap (`key, opt_names, ``{
	    string val = SCR::Read (add (.dns.named.value.options, key));
	    return [key, val];
	});
	if (options == nil)
	    options = $[];
	settings["options"] = options;
	y2milestone ("Read options: %1", options);

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (20);

	integer zonecount = size (zone_headers);
	integer done = 0;
	list zones = maplist (`z, zone_headers, ``{
	    string zonename = regexpsub (z, ".*\"(.*)\".*", "\\1");
	    string zonetype = SCR::Read (add (add (.dns.named.value, z),
		"type"));
	    string filename = SCR::Read (add (add (.dns.named.value, z),
		"file"));
	    if (filename != nil && regexpmatch (filename, "^\".*\"$"))
		filename = regexpsub (filename, "^\"(.*)\"$", "\\1");
	    map zd = $[];
	    if (zonetype == "master")
	    {
		zd = zoneRead (zonename, filename);
	    }
	    else if (zonetype == "slave" || zonetype == "stub")
	    {
		zd["masters"] = SCR::Read (add (add (.dns.named.value, z),
		    "masters"));
	    }
	    else
	    {
		// hint, forward, many others ...
		// not supported now
		// TODO in future
	    }
	    zd["file"] = filename;
	    zd["type"] = zonetype;
	    zd["zone"] = zonename;
	    done = done + 1;
	    Progress::Step (20 + done / zonecount);
	    sleep (500);
	    return zd;
	});
	y2milestone ("All zones: %1", zones);
	settings["zones"] = zones;

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (30);
	// progress stage
	Progress::Title(_("Finished"));

	modified = false;
	y2milestone ("Read settings %1", settings);
	return true;
    }

    /**
      * Write all dns-server settings
      * @return true on success
      */
     global define boolean Write() ``{

	/* DnsServer read dialog caption */
	string caption = _("Saving DNS Server Configuration");

	integer zones = size (settings["zones"]:[]);
	if (zones < 1)
	    zones = 1;

	integer steps = 3 * zones;

	Progress::New(caption, " ", steps, [
		// progress stage
		_("Save the zone files"),
		//  progress stage
		_("Save main configuration file"),
		// progress stage
		_("Restart DHCP server"),
	    ], [
		// progress step
		_("Saving the zone files..."),
		// progress step
		_("Saving main configuration file..."),
		// progress step
		_("Restarting DHCP server..."),
	    ],
	    ""
	);

	// write settings
	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (0);

	if (save_all)
	    modified = true;

	if (! modified)
	{
	    Progress::NextStageStep (3 * zones);
	    // progress stage
	    Progress::Title(_("Finished"));
	    return true;
	}


	date = getDate ();
	// be sure options are before zones
	SCR::Write (.dns.named.section.options, "");

	boolean ok = true;
	foreach (`z, settings["zones"]:[], ``{
	    ok = ok && saveZone (z);
	    Progress::NextStep ();
	});

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (zones);

	if (modified)
	    ok = ok &&saveGlobals ();
	if (! ok)
	    return false;

	SCR::Write (.sysconfig.named.NAMED_RUN_CHROOTED,
	    settings["chroot"]:false ? "yes" : "no");
	SCR::Write (.sysconfig.named, nil);

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (2 * zones);

	if (settings["start_service"]:false)
	{
	    integer ret = 0;
	    if (! write_only)
		ret = SCR::Execute (.target.bash, "/etc/init.d/named restart");
	    Runlevel::ServiceAdjust ("named", "enable");
	    if (0 != ret)
	    {
		// yes-no popup
		if (UI::YesNoPopup (_("Warning!

An error occurred while starting the DNS server.
Display the log and continue with configuration?
")))
		{
		    showLogPopup (nil);
		    return nil;
		}
		return false;
	    }
	}
	else
	{
	    if (! write_only)
		SCR::Execute (.target.bash, "/etc/init.d/named stop");
	    Runlevel::ServiceAdjust ("named", "disable");
	}

	if(UI::PollInput() == `abort) return false;
	Progress::NextStageStep (3 * zones);

	// progress stage
        Progress::Title(_("Finished"));

	return true;
    }

    /**
      * Get all dns-server settings from the first parameter
      * (For use by autoinstallation.)
      * @param imp_settings The YCP structure to be imported.
      * @return boolean True on success
      */
    global define boolean Import (map imp_settings) ``{
	settings = imp_settings;
	modified = true;
	zones_modified = true;
	save_all = true;
	return true;
    }

    /**
      * Dump the dns-server settings to a single map
      * (For use by autoinstallation.)
      * @return map Dumped settings (later acceptable by Import ())
      */
    global define map Export () ``{
	return settings;
    }

    /**
      * Create a textual summary and a list of unconfigured cards
      * @return list summary of the current configuration
      */
    global define list Summary() ``{
	map zone_types = $[
	    // type of zone to be used in summary
	    "master" : _("master"),
	    // type of zone to be used in summary
	    "slave" : _("slave"),
	    // type of zone to be used in summary
	    "stub" : _("stub"),
	    // type of zone to be used in summary
	    "hint" : _("hint"),
	    // type of zone to be used in summary
	    "forward" : _("forward"),
	];
	string start_string = settings["start_service"]:false
	    // summary string
	    ? _("The DNS server starts when booting the system.")
	    // summary string
	    : _("The DNS server does not start when booting the system.");
	list(list(string)) zones = maplist (`z, settings["zones"]:[], ``(
	    [z["zone"]:"", z["type"]:""]
	));
	zones = filter (`z, zones, ``(z[0]:"" != ""));
	list(string) zones_disp = maplist (`z, zones, ``(
	    z[1]:"" != ""
		? sformat ("%1 (%2)", z[0]:"", zone_types[(z[1]:"")]:(z[1]:""))
		: z[0]:""
	));
	// summary string, %1 is list of DNS zones (their names)
	string zones_string = sformat (_("Configured Zones: %1"),
	    mergestring (zones_disp, ", "));
	list ret = [ start_string, zones_string ];
	return ret;
    }

/* EOF */
}
