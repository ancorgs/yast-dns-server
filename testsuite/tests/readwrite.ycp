/**
 * File:
 *  rw.ycp
 *
 * Module:
 *  DHCP server configurator
 *
 * Summary:
 *  Read and write testsuite
 *
 * Authors:
 *  Jiri Srain <jsrain@suse.cz>
 *
 * $Id$
 *
 */

{
    include "testsuite.ycp";
    import "Progress";
    import "DnsServer";

    Progress::off ();

    map READ = $[
	// Runlevel
        "init": $[
            "scripts": $[
                "exists": true,
                "runlevel": $[
                    "dhcpd": $[
                        "start": [],
                        "stop":  [],
                        ],
                    ],
                // their contents is not important for ServiceAdjust
                "comment": $[
                    "dhcpd": $[],
                    ],
                ],
            ],

	"target" : $[
	    "stat" : $[],
	    "ycp" : $[],
	],
    ];
    map WRITE = $[];
    map EXEC = $[
	"target" : $[
	    "bash_output" : $[],
	],
    ];

    DUMP ("=======================");
    TEST (``(DnsServer::Read ()), [READ, WRITE, EXEC], 0);
    DUMP ("=======================");
    DnsServer::settings = $[
	"options":$[
	    "allow-query":"{ 127.0.0.1; }",
	     "directory":"\"/var/lib/named\"",
	], "zones":[
	    $[
		"a":$["zealot.suse.cz.":["127.0.0.1"]],
		"cname":$[],
		"expiry":"6W",
		"file":"/var/lib/named/localhost.zone",
		"mail":"root",
		"minimum":"1 W",
		"mx":[],
		"ns":["@"],
		"ptr":$[],
		"refresh":"2D",
		"retry":"4H",
		"server":"zealot.suse.cz.",
		"ttl":"1W",
		"type":"master",
		"zone":"localhost",
	    ],
	]
    ];
    TEST (``(DnsServer::Write ()), [READ, WRITE, EXEC], 0);
    DUMP ("=======================");
}
