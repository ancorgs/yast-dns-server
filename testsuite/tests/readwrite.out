Dump	=======================
Execute	.target.bash_output "/bin/hostname -f" $[]
Log	Hostname detection failed, using localhost
Execute	.target.bash "rpm -q bind8" 0
Execute	.target.bash "rpm -q bind9" 0
Read	.init.scripts.exists "named" true
Read	.init.scripts.runlevel "named" $["named":$["start":[], "stop":[]]]
Read	.sysconfig.named.NAMED_RUN_CHROOTED 0
Dir	.dns.named.section: []
Dir	.dns.named.value.options: []
Return	true
Dump	=======================
Execute	.target.bash_output "date '+%Y%m%d'" $[]
Write	.dns.named.section.options "" true
Write	.dns.named.section."zone \"localhost\" in" "" true
Dir	.dns.named.value.options: []
Write	.dns.zone ["/var/lib/named/localhost.zone", $["TTL":"1W", "records":[$["key":"", "type":"A", "value":"127.0.0.1"], $["key":"", "type":"NS", "value":"@"]], "soa":$["expiry":"6W", "mail":"root", "minimum":"1W", "refresh":"2D", "retry":"4H", "serial":"00", "server":"@", "zone":"@"]]] true
Write	.dns.named.value."zone \"localhost\" in"."file" "\"localhost.zone\"" true
Write	.dns.named.value."zone \"localhost\" in"."type" "master" true
Dir	.dns.named.section: []
Dir	.dns.named.value.options: []
Write	.dns.named.value.options."allow-query" "{ 127.0.0.1; }" true
Write	.dns.named.value.options."directory" "\"/var/lib/named\"" true
Write	.dns.named nil true
Write	.sysconfig.named.NAMED_RUN_CHROOTED "no" true
Write	.sysconfig.named nil true
Execute	.target.bash "/etc/init.d/named stop" 0
Read	.init.scripts.exists "named" true
Read	.init.scripts.exists "named" true
Read	.init.scripts.runlevel "named" $["named":$["start":[], "stop":[]]]
Read	.init.scripts.comment "named" $["named":$[]]
Return	true
Dump	=======================
